test/e2e/secondary_network_ipam_test.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"antrea.io/antrea/pkg/agent/config"
+	crdv1alpha2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	agentconfig "antrea.io/antrea/pkg/config/agent"
+	controllerconfig "antrea.io/antrea/pkg/config/controller"
+)
+
+var (
+	testIPPoolv4 = &crdv1alpha2.IPPool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-ippool-ipv4-1",
+		},
+		Spec: crdv1alpha2.IPPoolSpec{
+			IPVersion: crdv1alpha2.IPv4,
+			IPRanges: []crdv1alpha2.SubnetIPRange{{IPRange: crdv1alpha2.IPRange{
+				CIDR:  "",
+				Start: "10.123.1.101",
+				End:   "10.123.1.200",
+			},
+				SubnetInfo: crdv1alpha2.SubnetInfo{
+					Gateway:      "10.123.1.254",
+					PrefixLength: 24,
+				}}},
+		},
+	}
+
+	cniCmd = "/opt/cni/bin/antrea"
+
+	cniEnvs = map[string]string{
+		"CNI_CONTAINERID": "test-container-id",
+		"CNI_NETNS":       "/var/run/netns/test-netns",
+		"CNI_PATH":        "/opt/cni/bin",
+		"CNI_ARGS":        "K8S_POD_NAMESPACE=test-namespace;K8S_POD_NAME=test-pod",
+	}
+
+	cniNetworkConfig = `{
+    "cniVersion": "0.3.0",
+    "name": "test",
+    "type": "test-cni",
+    "keyA": ["some more", "plugin specific", "configuration"],
+    "ipam": {
+        "type": "antrea",
+        "ippools": [ "test-ippool-ipv4-1" ],
```

Comment:
Just a question, where does this parameter come from?

Reply:
I created the IPPool named "test-ippool-ipv4-1" in this test. Check line 30 - 46.

Reply:
To clarify this question, I want to ask how this parameter passed in this cniNetworkConfig.
From the PR description, we have a NetworkAttachmentDefinition, and I know some special config will be passed in CNI call, but I think we still need some other configuration to make the whole pipeline work.

Another question: do we want to cover dns/routes support in this PR or in Antrea?

Reply:
In the real usage with Multus, Multus will pass the whole CNI config json in NetworkAttachmentDefinition to the main CNI plugin (e.g. MACVLAN), which then passes the json to the IPAM plugin (just like how Antrea invokes node-ipam plugin). The behavior of the latter part (from main CNI to IPAM CNI) is defined in the CNI spec.

I have not heard about any request to support dns/routes for the primary CNI. Added them for secondary network also for feature parity with Whereabouts.

Reply:
Got it. Thanks.